build: Recompile everything on `.config` changes (workaround)

skuenzer opened the pull request:
### Description of changes

Unfortunately, our build system is still not able to re-trigger the compilation if only the command line of a compilation unit changed due to a configuration change. This can result in broken builds if the user did not call `make clean` in between.

This commit introduces a workaround that attempts to improve the user experience by triggering a recompile when the configuration is changed. This makes this cleaning step unecessary.

It is intended that commit "build: Recompile on `.config` changes (workaround)" gets reverted as soon as the build system can properly detect changes to the compiler command line.

This PR depends on PR #1070.
This is a workaround for issue #47.

skuenzer committed:
build: Link-time dependencies for libraries (`*_LDEPS`)

For each `.o` library, additional dependencies that trigger relinking
can be specified in `Makefile.uk` with `*_LDEPS`:
```
<LIBNAME>_LDEPS   += <dependency>
<LIBNAME>_LDEPS-y += <dependency>
```

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
build: Restructure common dependencies for preprule_*

Restructures the preprocess rules to put common dependencies in a single
place. The advantage is that these dependencies are now automatically
applied for externally defined prepare rules.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
build: Fix rule that copies `.config` to `build/config`

The rule that copies the current `.config` to the build directory was
intended, among other things, to detect configuration changes. However, the
corresponding rule always copies the file, causing to update the timestamp
on every make call. This commit fixes that rule by only copying the file
when there have been configuration updates.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
build: Recompile on `.config` changes (workaround)

Our build system is still not able to re-trigger the compilation if only
the command line of a compilation unit changed due to a configuration
change. This resulted in broken builds if the user did not call `make
clean` in between.
This commit introduces a workaround that attempts to improve the user
experience by triggering a recompile when the configuration is changed.
This makes this cleaning step unecessary.

It is intended that this commit gets reverted as soon as the build system
can properly detect changes to the compiler command line.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
scripts: configupdate: Only notify on architecture change

The `configupdate` script tells the user when `make properclean` should be
called. Since compilation is retriggered on command line changes, we can
limit the cases where cleanup is suggested to architecture changes. This is
done to ensure that the architecture is also considered for fetch and
prepare targets.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

StefanJum reviewed (approved):
This works fine, thank you. I'll add the tag after #1070 gets in and this gets rebased.

jschlumpp commented on support/build/Makefile.rules:
```diff
@@ -813,7 +813,7 @@ $(error buildrule_$(call fileext,$(strip $(2))) is not defined: Failed to instal
 
 # Default buildrule_* dependencies
-$(4): $(call qstrip,$($(call vprefix_src,$(1),$(2),$(3),CDEPS)) $($(call vprefix_src,$(1),$(2),$(3),CDEPS-y))) | preprocess
+$(4): $(call qstrip,$($(call vprefix_src,$(1),$(2),$(3),CDEPS)) $($(call vprefix_src,$(1),$(2),$(3),CDEPS-y))) $(UK_CONFIG_OUT) | preprocess
```
I think this should also include a comment about the workaround (Same for the next hunk)

michpappas replied:
+1

michpappas replied:
+1

skuenzer replied:
I added 👍🏻

michpappas commented on support/scripts/configupdate:
```diff
@@ -25,18 +25,12 @@ NL=$'\n'
 # Change of architecture?
 _comparesym "CONFIG_ARCH_" "${UK_CONFIG}" "${UK_CONFIG_OLD}" || UPDATES+="*** - CPU architecture changed${NL}"
 
-# Change of CPU options?
```
Since the full rebuild is only a transient measure, won't these be required once we move to a proper fix?

skuenzer replied:
Even after transiting, we should be able to properly handle re-compilation on updated compile command line. This means that CPU options and optimization should be covered. Actually, also the architecture should be covered (since we call a different compiler). I thought of even removing the architecture line but I don't know if we should keep it due external libraries. I am currently imagining a library that downloads different sources depending on the architecture. 🤔

skuenzer committed:
build: Clean-up and removing unused targets, like `config`

This is a small clean-up commit: it removes the unused targets `config`
and `%_defconfig`. Additionally, references to non-existing targets, like
`outputmakefile`, `silentoldconfig`, `list-defconfigs` are removed.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

michpappas reviewed (approved):
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
